### PR TITLE
[SPARK-34598][SQL] RewritePredicateSubquery Rule must not update Filters without subqueries

### PR DESCRIPTION
**Pull request by Swinky**
### What changes were proposed in this pull request?
RewritePredicateSubquery Optimizer Rule must not update Filters without subqueries.

Following is one such example.

```
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.RewritePredicateSubquery ===
 Project [a#0]                                                        Project [a#0]
!+- Filter (((a#0 > 1) OR (b#1 > 2)) AND ((c#2 > 1) AND (d#3 > 2)))   +- Filter ((((a#0 > 1) OR (b#1 > 2)) AND (c#2 > 1)) AND (d#3 > 2))
    +- LocalRelation <empty>, [a#0, b#1, c#2, d#3]                       +- LocalRelation <empty>, [a#0, b#1, c#2, d#3]
```

### Why are the changes needed?
minor change.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing UTs pass.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 1]
```diff
@@ -98,55 +98,59 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
         splitConjunctivePredicates(condition)
           .partition(SubqueryExpression.hasInOrCorrelatedExistsSubquery)
 
-      // Construct the pruned filter condition.
-      val newFilter: LogicalPlan = withoutSubquery match {
-        case Nil => child
-        case conditions => Filter(conditions.reduce(And), child)
```
Comment (Swinky) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:104]: @cloud-fan do we want to update `conditions.reduce(And)` as well?
similar to the review suggestion [here](https://github.com/apache/spark/pull/31318#discussion_r579979610)

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
Can you add a test to demonstrate the problem?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 2]
```diff
@@ -98,55 +98,59 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
...
-          val (joinCond, outerPlan) = rewriteExistentialExpr(inConditions ++ conditions, p)
-          Join(outerPlan, newSub, LeftSemi, joinCond, JoinHint.NONE)
-        case (p, Not(InSubquery(values, ListQuery(sub, conditions, _, _)))) =>
-          // This is a NULL-aware (left) anti join (NAAJ) e.g. col NOT IN expr
-          // Construct the condition. A NULL in one of the conditions is regarded as a positive
-          // result; such a row will be filtered out by the Anti-Join operator.
-
-          // Note that will almost certainly be planned as a Broadcast Nested Loop join.
-          // Use EXISTS if performance matters to you.
+      withSubquery match {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:101]: Why not just check if the filter condition has correlated in/exists subquery?
```scala
case f @ Filter(condition, child)
    if SubqueryExpression.hasInOrCorrelatedExistsSubquery(condition) =>
```

**Comment by HyukjinKwon** [comment]
cc @maryannxue FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala (hunk 2 above)
Comment (Swinky) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala]: done, thanks @allisonwang-db

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteSubquerySuite.scala [hunk 3]
```diff
@@ -67,4 +68,12 @@ class RewriteSubquerySuite extends PlanTest {
 
     comparePlans(optimized, correctAnswer)
   }
+
+  test("SPARK-34598: Filters without subquery must not be modified by RewritePredicateSubquery") {
+    val relation = LocalRelation('a.int, 'b.int, 'c.int, 'd.int)
+    val query = relation.where(('a === 1 || 'b === 2) && ('c === 3 && 'd === 4)).select('a)
+    val tracker = new QueryPlanningTracker
+    Optimize.executeAndTrack(query.analyze, tracker)
+    assert(tracker.rules.get(RewritePredicateSubquery.ruleName).get.numEffectiveInvocations == 0)
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteSubquerySuite.scala:77]: nit: `assert(tracker.rules(RewritePredicateSubquery.ruleName).numEffectiveInvocations == 0)`

**Review by maropu — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 4]
```diff
@@ -93,7 +93,8 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
   }
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
-    case Filter(condition, child) =>
+    case Filter(condition, child)
+      if SubqueryExpression.hasInOrCorrelatedExistsSubquery(condition) =>
```
Comment (maryannxue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:97]: Nit: can we pay attention to the indent here? should go 2 spaces further in.

**Pull request closed without merging**